### ISSUE TITLE
Bug: v4 OIDC implementation - login won't work

### ISSUE THREAD

**Issue by brunostjohn** [dicedtomatoreal/typex#637]
### What happened?

When signing in on a new machine, OIDC auth fails returning `{"code":401,"message":"invalid session"}`.

Flow:
1. Click on "Sign in with OIDC"
2. Authenticate with IDP
3. See above JSON returned

Symptoms:
- No session is appended to the session array in the database (Postgres).

Fixed by:
- Add an empty string to the text array (so the Prisma query matches).
- Remove user's oauth provider row in the database.
- Authenticate

### Version

other (provide version in additional info)

### What browser(s) are you seeing the problem on?

Chromium-based (Chrome, Edge, Brave, Opera, mobile chrome/chromium based, etc)

### Zipline Logs

```shell
[2024-12-05T16:49:01 INFO   db] connecting to database postgres://zipline:[password]@postgres-cluster-rw.databases.svc.cluster.local:5432/zipline?sslmode=disable
[2024-12-05T16:49:01 INFO   server] starting zipline mode="production" version="4.0.0-dev+1"
[2024-12-05T16:49:01 INFO   server] reading settings...
[2024-12-05T16:49:07 INFO   server] server started hostname="0.0.0.0" port=3000
[2024-12-05T16:49:36 ERROR  tasks::thumbnail-0] failed to generate thumbnail err="Cannot find ffmpeg"
[2024-12-05T16:49:38 ERROR  tasks::thumbnail-1] failed to generate thumbnail err="Cannot find ffmpeg"
[2024-12-05T16:50:21 ERROR  server] UnauthorizedError: not logged in
[2024-12-05T16:50:21 ERROR  server] NotFoundError: Not Found
[2024-12-05T16:50:26 ERROR  server] UnauthorizedError: not logged in
[2024-12-05T16:50:29 ERROR  server] UnauthorizedError: not logged in
```

### Browser Logs

```shell
GET [domain]/[oauth api route in pages folder] 401 Unauthorized
```

### Additional Info

Version: `ghcr.io/diced/zipline:v4-5f9762f`

My ideas for why this could be happening.
In the `withOauth` middleware, a user is searched for by a session id:
```ts
    const user = await prisma.user.findFirst({
      where: {
        sessions: {
          has: session.sessionId ?? '',
        },
      },
      include: {
        oauthProviders: true,
      },
    });
```
This will fail for a user that has not previously signed in on a given machine as they do not have a session there yet. The sign in button, for some reason, always returns the `link` state in the URL. This means the server will always attempt to "link" an account with an OAuth provider. This means that even if a user has a session, going through the sign in route will result in attempting to re-link the account.

Since the `withOauth` middleware runs after the oidc route and that route saves no session, it's as if the user was unauthenticated from the get go and an `invalid session` gets returned here:
```ts
    if (urlState === 'link') {
      if (!user) return res.unauthorized('invalid session');

      if (findProvider(provider, user.oauthProviders))
        return res.badRequest('This account is already linked to this provider');
```
The first error being returned when signing in on a new machine, the second being returned with a valid session but an already linked account. This if seems to evaluate as true during a normal sign in.

I'm a bit badly rested recently, so I might've been wrong somewhere here. In that case, I'm sorry. Hope this helps!

In short, what I think is happening:
- The IDP authenticates the user
- A session token isn't set when that happens
- The withOIDC middleware attempts to find the user by a session found in the browser cookies
- No session is set though
- Auth fail

To repro:
- Create new account
- Link OIDC IDP
- Sign out
- Clear cookies +local/session storage
- Attempt OIDC sign in
- Bob's your uncle

**Comment by diced** [comment]
Looks like the issue was that every time /api/auth/oauth/provider was accessed it would think that you are trying to link your signed in account (in this case, you aren't trying to link, you are trying to register) which would cause this error saying invalid session since there wasn't a session in the first place. Was just a small bug while I was implementing sessions and is fixed in 05a4e08fba9045357dd845ec77eeba6110ee588d
